File: src/optimization/bmu.jl
"""
Best Matching Unit (BMU) finding for BasicDBGSOM.
"""

# ============================================================================
# Single Sample BMU
# ============================================================================

"""
    find_bmu(topo, x) -> (position, distance)

Find the Best Matching Unit for input vector x.
Automatically handles NaN values in input by using scaled partial distances.

# Returns
- `position::Tuple{Int,Int}`: Grid position of BMU
- `distance::T`: Euclidean distance to BMU
"""
function find_bmu(topo::SOMTopology{T}, x::AbstractVector{T}) where T
    best_pos = nothing
    best_dist_sq = typemax(T)

    # Check once if input has NaN
    x_has_nan = any(isnan, x)

    for (pos, neuron) in topo.neurons
        dist_sq = if x_has_nan
            euclidean_squared_nan(neuron.weights, x)
        else
            euclidean_squared(neuron.weights, x)
        end
        if dist_sq < best_dist_sq
            best_dist_sq = dist_sq
            best_pos = pos
        end
    end

    return best_pos, sqrt(best_dist_sq)
end

find_bmu(som::DBGSOM{T}, x::AbstractVector{T}) where T = find_bmu(som.topology, x)

"""
    find_bmu_idx(positions, D_row) -> Int

Find BMU index from precomputed distance row.
"""
function find_bmu_idx(D_row::AbstractVector{T}) where T
    best_idx = 1
    best_dist = D_row[1]

    @inbounds for i in 2:length(D_row)
        if D_row[i] < best_dist
            best_dist = D_row[i]
            best_idx = i
        end
    end

    return best_idx
end

# ============================================================================
# Batch BMU Finding
# ============================================================================

"""
    find_bmus(topo, X) -> Vector{Tuple{Int,Int}}

Find BMUs for all samples in X.
Automatically handles NaN values by using scaled partial distances.

# Arguments
- `topo::SOMTopology{T}`: Topology with neurons
- `X::AbstractMatrix{T}`: Input data (input_dim × n_samples)

# Returns
Vector of BMU positions for each sample.
"""
function find_bmus(topo::SOMTopology{T}, X::AbstractMatrix{T}) where T
    n_samples = size(X, 2)
    pos_list, W = get_positions_and_weights(topo)

    # Compute all pairwise distances (auto-dispatches based on NaN presence)
    D = pairwise_distances_squared_auto(Matrix(X), W)

    # Find argmin for each sample
    bmus = Vector{Tuple{Int,Int}}(undef, n_samples)

    @inbounds for i in 1:n_samples
        idx = find_bmu_idx(view(D, i, :))
        bmus[i] = pos_list[idx]
    end

    return bmus
end

find_bmus(som::DBGSOM{T}, X::AbstractMatrix{T}) where T = find_bmus(som.topology, X)

"""
    find_bmus(topo, X, miss_info) -> Vector{Tuple{Int,Int}}

Find BMUs using missingness-weighted distances when NaN data is present.
Falls back to standard distance when miss_info is nothing or data has no NaN.
"""
function find_bmus(topo::SOMTopology{T}, X::AbstractMatrix{T},
                   miss_info::MissingnessInfo{T}) where T
    n_samples = size(X, 2)
    pos_list, W = get_positions_and_weights(topo)

    D = if has_nan(X)
        pairwise_distances_squared_nan_weighted(Matrix(X), W, miss_info)
    else
        pairwise_distances_squared(Matrix(X), W)
    end

    bmus = Vector{Tuple{Int,Int}}(undef, n_samples)
    @inbounds for i in 1:n_samples
        idx = find_bmu_idx(view(D, i, :))
        bmus[i] = pos_list[idx]
    end

    return bmus
end

"""
    dispatch_find_bmus(topo, X, miss_info) -> Vector{Tuple{Int,Int}}

Centralized NaN-aware BMU dispatch. Uses missingness-weighted distances when
`miss_info` is provided and data contains NaN; otherwise uses standard BLAS path.
"""
function dispatch_find_bmus(topo::SOMTopology{T}, X::AbstractMatrix{T},
                            miss_info::Union{Nothing, MissingnessInfo{T}}) where T
    if miss_info !== nothing
        find_bmus(topo, X, miss_info)
    else
        find_bmus(topo, X)
    end
end

"""
    find_bmus_with_distances(topo, X) -> (Vector{Tuple{Int,Int}}, Vector{T})

Find BMUs and their distances for all samples.
Automatically handles NaN values by using scaled partial distances.

# Returns
- `bmus::Vector{Tuple{Int,Int}}`: BMU positions
- `distances::Vector{T}`: Euclidean distances to BMUs
"""
function find_bmus_with_distances(topo::SOMTopology{T}, X::AbstractMatrix{T}) where T
    n_samples = size(X, 2)
    pos_list, W = get_positions_and_weights(topo)

    # Compute all pairwise distances (auto-dispatches based on NaN presence)
    D = pairwise_distances_squared_auto(Matrix(X), W)

    bmus = Vector{Tuple{Int,Int}}(undef, n_samples)
    distances = Vector{T}(undef, n_samples)

    @inbounds for i in 1:n_samples
        idx = find_bmu_idx(view(D, i, :))
        bmus[i] = pos_list[idx]
        distances[i] = sqrt(D[i, idx])
    end

    return bmus, distances
end

find_bmus_with_distances(som::DBGSOM{T}, X::AbstractMatrix{T}) where T =
    find_bmus_with_distances(som.topology, X)

"""
    find_bmus_with_distances(topo, X, miss_info) -> (Vector{Tuple{Int,Int}}, Vector{T})

Find BMUs and distances using missingness-weighted metric when NaN data is present.
Falls back to standard BLAS distance when data has no NaN.
"""
function find_bmus_with_distances(topo::SOMTopology{T}, X::AbstractMatrix{T},
                                  miss_info::MissingnessInfo{T}) where T
    n_samples = size(X, 2)
    pos_list, W = get_positions_and_weights(topo)

    D = if has_nan(X)
        pairwise_distances_squared_nan_weighted(Matrix(X), W, miss_info)
    else
        pairwise_distances_squared(Matrix(X), W)
    end

    bmus = Vector{Tuple{Int,Int}}(undef, n_samples)
    distances = Vector{T}(undef, n_samples)

    @inbounds for i in 1:n_samples
        idx = find_bmu_idx(view(D, i, :))
        bmus[i] = pos_list[idx]
        distances[i] = sqrt(D[i, idx])
    end

    return bmus, distances
end

File: R/bmu.R
# bmu.R - Best Matching Unit finding for DBGSOM

#' Find BMU index from a precomputed distance row
#' @param d_row numeric vector of squared distances
#' @return integer index of minimum
find_bmu_idx <- function(d_row) {
  which.min(d_row)
}

#' Find BMUs for all samples using precomputed distance matrix
#'
#' @param neurons named list of neurons
#' @param X matrix (input_dim x n_samples)
#' @return character vector of position keys (one per sample)
find_bmus <- function(neurons, X) {
  wm <- get_weights_matrix(neurons)
  D <- pairwise_dist_sq_auto(X, wm$W)
  n_samples <- ncol(X)
  bmu_keys <- character(n_samples)
  for (i in seq_len(n_samples)) {
    idx <- find_bmu_idx(D[i, ])
    bmu_keys[i] <- wm$pos_keys[idx]
  }
  bmu_keys
}

#' Find BMUs using missingness-weighted distances
#'
#' @param neurons named list of neurons
#' @param X matrix (input_dim x n_samples)
#' @param miss_info missingness info list (or NULL)
#' @return character vector of position keys
find_bmus_weighted <- function(neurons, X, miss_info) {
  wm <- get_weights_matrix(neurons)
  if (!is.null(miss_info) && has_na(X)) {
    D <- pairwise_dist_sq_na_weighted(X, wm$W, miss_info)
  } else {
    D <- pairwise_dist_sq_auto(X, wm$W)
  }
  n_samples <- ncol(X)
  bmu_keys <- character(n_samples)
  for (i in seq_len(n_samples)) {
    bmu_keys[i] <- wm$pos_keys[find_bmu_idx(D[i, ])]
  }
  bmu_keys
}

#' Dispatch BMU finding: use weighted if miss_info available, else standard
#' @param neurons named list of neurons
#' @param X matrix (input_dim x n_samples)
#' @param miss_info missingness info list or NULL
#' @return character vector of position keys
dispatch_find_bmus <- function(neurons, X, miss_info) {
  if (!is.null(miss_info)) {
    find_bmus_weighted(neurons, X, miss_info)
  } else {
    find_bmus(neurons, X)
  }
}

#' Find BMUs and their distances for all samples
#' @param neurons named list of neurons
#' @param X matrix (input_dim x n_samples)
#' @return list with bmu_keys (character) and distances (numeric)
find_bmus_with_distances <- function(neurons, X) {
  wm <- get_weights_matrix(neurons)
  D <- pairwise_dist_sq_auto(X, wm$W)
  n_samples <- ncol(X)
  bmu_keys <- character(n_samples)
  distances <- numeric(n_samples)
  for (i in seq_len(n_samples)) {
    idx <- find_bmu_idx(D[i, ])
    bmu_keys[i] <- wm$pos_keys[idx]
    distances[i] <- sqrt(D[i, idx])
  }
  list(bmu_keys = bmu_keys, distances = distances)
}

#' Find first and second BMUs for all samples (for topographic error)
#' @param neurons named list of neurons
#' @param X matrix (input_dim x n_samples)
#' @return list with bmu1_keys, bmu2_keys, dist1s, dist2s
find_two_bmus <- function(neurons, X) {
  wm <- get_weights_matrix(neurons)
  D <- pairwise_dist_sq_auto(X, wm$W)
  n_samples <- ncol(X)
  n_neurons <- ncol(D)

  bmu1_keys <- character(n_samples)
  bmu2_keys <- character(n_samples)
  dist1s <- numeric(n_samples)
  dist2s <- numeric(n_samples)

  for (i in seq_len(n_samples)) {
    row <- D[i, ]
    # First minimum
    idx1 <- which.min(row)
    # Second minimum (exclude first)
    row2 <- row
    row2[idx1] <- Inf
    idx2 <- which.min(row2)

    bmu1_keys[i] <- wm$pos_keys[idx1]
    bmu2_keys[i] <- wm$pos_keys[idx2]
    dist1s[i] <- sqrt(row[idx1])
    dist2s[i] <- sqrt(row[idx2])
  }

  list(bmu1_keys = bmu1_keys, bmu2_keys = bmu2_keys,
       dist1s = dist1s, dist2s = dist2s)
}
